pkg/ha/ha_service.go

```diff
@@ -103,5 +105,70 @@ func (ha *HAService) Start(ctx context.Context) error {
 		}
 	}
 
-	return nil
+	<-ctx.Done()
+	err := ha.cleanUp()
+	if err == nil {
+		err = ctx.Err()
+	}
+	return err
+}
+
+// cleanUp is executed upon ending leadership. Its purpose is to remove the Endpoints object created upon acquiring
+// leadership.
+func (ha *HAService) cleanUp() error {
+	// Use our own context. This function executes when the main application context is closed.
+	// Also, try to finish before a potential 15 seconds termination grace timeout.
+	ctx, cancel := context.WithTimeout(context.Background(), 14*time.Second)
+	defer cancel()
+	seedClient := ha.manager.GetClient()
+
+	attempt := 0
+	var err error
+	for {
+		endpoints := corev1.Endpoints{}
+		err = seedClient.Get(ctx, client.ObjectKey{Namespace: ha.namespace, Name: app.Name}, &endpoints)
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				ha.log.V(app.VerbosityVerbose).Info("The endpoints object cleanup succeeded: the object was missing")
+				return nil
+			}
+
+			ha.log.V(app.VerbosityInfo).Info("Failed to retrieve the endpoints object", "error", err.Error())
+		} else {
+			// Avoid data race. We don't want to delete the endpoint if it is sending traffic to a replica other than this one.
+			isEndpointStillPointingToOurReplica :=
```

Comment:
We could be reading a cached object here that still points to this replica. Or we could fetch the object right before it is changed by the new leader.